### PR TITLE
SOC-1161 Fix bolding of watchlist items for Wall

### PR DESCRIPTION
**Pull request by garthwebb**
This is the original ticket:

https://wikia-inc.atlassian.net/browse/VOLDEV-159

This change was made in a non-Wikia repo so this ticket is used as the merge request into the Wikia repo:

https://wikia-inc.atlassian.net/browse/SOC-1611

### CODE REVIEW

File: extensions/wikia/Wall/WallThread.class.php [hunk 1]
```diff
@@ -117,7 +117,8 @@ private function getReplyIdsFromDB( $dbr, $afterId = null ) {
 	}
 
 	private function loadReplyIdsFromDB( $master = false ) {
-		if ( empty( Title::newFromId( $this->mThreadId ) ) ) {
+		// VOLDEV-3 change: fix "Can't use function return value in write context" fatal
+		if ( Title::newFromId( $this->mThreadId ) == null ) {
```
Comment (kvas-damian) [inline extensions/wikia/Wall/WallThread.class.php]: Please use strict comparison.

File: extensions/wikia/ArticleComments/classes/ArticleComment.class.php [hunk 2]
```diff
@@ -1114,19 +1114,19 @@ static public function addArticlePageToWatchlist( $comment ) {
 	 * @return Bool true -- because it's a hook
 	 */
 	static public function watchlistNotify( RecentChange &$oRC ) {
-		global $wgEnableGroupedArticleCommentsRC;
-		wfProfileIn( __METHOD__ );
+		$app = F::app();
```
Comment (garthwebb) [inline extensions/wikia/ArticleComments/classes/ArticleComment.class.php]: Given `$app` isn't used directly for anything else but `wg`, this could be:

``` php
$wg = F::app()->wg;
```

and we could reduce line length on subsequent lines.

File: extensions/wikia/Wall/WallHooksHelper.class.php [hunk 3]
```diff
@@ -686,8 +686,33 @@ static public function onArticleDelete( $article, &$user, &$reason, &$error ) {
...
-		wfProfileIn( __METHOD__ );
+	/**
+	 * Adjusting Recent Changes for Wall
+	 * Change RC db key to affect Watchlist behavior
+	 * @author TK-999 <https://github.com/TK-999>
+	 * @param RecentChange $rc
+	 * @return bool true because it is a hook
+	 */
+	static public function onRecentChangeBeforeSave( RecentChange &$rc ) {
+		$app = F::app();
```
Comment (garthwebb) [inline extensions/wikia/Wall/WallHooksHelper.class.php]: Same comment as above regarding `F::app()->wg`

File: extensions/wikia/Wall/WallHooksHelper.class.php [hunk 4]
```diff
@@ -1458,9 +1496,14 @@ static public function onArticleSaveComplete( &$article, &$user, $text, $summary
...
-		$app = F::app();
-		if ( in_array( MWNamespace::getSubject( $title->getNamespace() ), $app->wg->WallNS ) || $title->getNamespace() == NS_USER_WALL_MESSAGE_GREETING ) {
+	/**
+	 * Don't send email for Wall changes
+	 * @param User $editor
+	 * @param Title $title
+	 * @return bool
+	 */
+	static public function onAllowNotifyOnPageChange( User $editor, Title $title ) {
+		if ( in_array( MWNamespace::getSubject( $title->getNamespace() ), F::app()->wg->WallNS ) || $title->getNamespace() == NS_USER_WALL_MESSAGE_GREETING ) {
```
Comment (garthwebb) [inline extensions/wikia/Wall/WallHooksHelper.class.php]: This line is getting a bit long (I think we prefer under 120 chars).  This could probably be made more clear as two conditionals:

``` php
if ( in_array( MWNamespace::getSubject( $title->getNamespace() ), F::app()->wg->WallNS ) {
    return false;
}

if ( $title->getNamespace() == NS_USER_WALL_MESSAGE_GREETING ) {
    return false;
}
```

**Comment by garthwebb** [comment]
@TK-999 left some comments for you here!

**Comment by mszabo-wikia** [comment]
@garthwebb Sorry about the delay, had exams :) See https://github.com/Wikia/app/pull/9602

File: extensions/wikia/Wall/WallHooksHelper.class.php [hunk 5]
```diff
@@ -686,8 +686,33 @@ static public function onArticleDelete( $article, &$user, &$reason, &$error ) {
...
 	}
 
-	static public function onRecentChangeSave( $recentChange ) {
-		wfProfileIn( __METHOD__ );
+	/**
+	 * Adjusting Recent Changes for Wall
+	 * Change RC db key to affect Watchlist behavior
+	 * @author TK-999 <https://github.com/TK-999>
+	 * @param RecentChange $rc
+	 * @return bool true because it is a hook
```
Comment (bkoval) [inline extensions/wikia/Wall/WallHooksHelper.class.php]: I can see capital "Bool" here - https://github.com/Wikia/app/pull/9001/files#diff-49ba2b005c744aa46f66a4717763c10cR1114 and this one is all small letters - @garthwebb @kvas-damian Which is the PHPDoc convention we should follow?

Comment (garthwebb) [inline extensions/wikia/Wall/WallHooksHelper.class.php]: I'm not sure we have any guidelines for this so I would suggest being consistent within the file and/or changeset.

Comment (mszabo-wikia) [inline extensions/wikia/Wall/WallHooksHelper.class.php]: @garthwebb @bkoval updated new PR

File: extensions/wikia/Wall/WallHooksHelper.class.php [hunk 6]
```diff
@@ -2339,8 +2339,11 @@ public static function onArticleCommentGetSquidURLs( $title, &$urls ) {
 	 */
 	private static function shouldBoldenFollowedLink( User $user, Title $title, WallMessage $wm ) {
 		$isWatchlist = $title->isSpecial( 'Watchlist' );
```
Comment (garthwebb) [inline extensions/wikia/Wall/WallHooksHelper.class.php]: Its a little weird that we're using "not Watchlist" (line 2340) as a test for "is RecentChanges".  I know this code has had a lot of updates, but it might be better to test for what we're looking for, e.g.:

``` php
$isRecentChanges = $title->isSpecial( 'RecentChanges' );
...
&& ( $isRecentChanges ... )
```

File: extensions/wikia/Wall/WallHooksHelper.class.php [hunk 7]
```diff
@@ -2339,8 +2339,11 @@ public static function onArticleCommentGetSquidURLs( $title, &$urls ) {
...
-		return ( F::app()->wg->ShowUpdatedMarker && !$user->isAnon()
+		return ( !$user->isAnon()
+			// Is this user watching this comment, or the thread, or the wall/board?
 			&& ( $wm->isWatched( $user ) || $wm->isWallWatched( $user ) || $wm->isWallOwner( $user ) )
+			// If yes, are we on RecentChanges, or was the page updated?
+			&& ( !$isWatchlist || $wm->getTitle()->getNotificationTimestamp( $user ) || $wm->getArticleTitle()->getNotificationTimestamp( $user ) )
```
Comment (garthwebb) [inline extensions/wikia/Wall/WallHooksHelper.class.php]: I think this might be easier to follow if each `&&` set were broken out into separate conditionals, e.g. from:

``` php
return ( A && B && C );
```

to 

``` php
if ( !A ) {
  return false;
}
if ( !B ) {
  return false;
}
if ( !C ) {
  return false;
}
return true;
```

Additionally, anything loaded from `$wm`, e.g. the `$wm->getTitle()` from `$wm->getTitle()->getNotificationTimestamp( $user )` should be tested first.  Wall code is notorious for having stale IDs, so its very likely that `$wm->getTitle()` could return `null` instead of a title and cause a runtime error.  If you follow my first suggestion you could do this easily by noting that in the last `C` conditional can be written as:

``` php
if ( !( C1 || C2 || C3) ) {
  return false;
}
return true;
```

and is the same as;

``` php
if ( C1 ) {
  return true;
}
if ( C2 ) {
  return true;
}
if ( C3 ) {
  return true;
}
return false
```

Expanding it this way would let you add the test for null objects easily and clearly:

``` php
if ( $wm->getTitle() && $wm->getTitle()->getNotificationTimestamp( $user ) ) {
  return true;
}
```

This makes the method longer, but I think ultimately more readable.

**Comment by garthwebb** [comment]
@TK-999, Hi, I left some detailed comments on the last commit, https://github.com/Wikia/app/pull/9001/commits/4132c07a68d57ef8a2906107ff712b4e65e14218

There's a lot there, but I think the most important thing is testing to see if the title or article object is null before calling a function on it.

File: extensions/wikia/Wall/WallBaseController.class.php [hunk 8]
```diff
@@ -70,7 +70,9 @@ public function thread() {
 		if ( !empty( $title ) && $title->exists() && in_array( MWNamespace::getSubject( $title->getNamespace() ), $this->app->wg->WallNS ) ) {
 			$wallMessage = WallMessage::newFromTitle( $title );
 			$wallMessage->load();
-			$this->app->wg->Out->setPageTitle( $wallMessage->getMetaTitle() );
+			// VOLDEV-3: fix Thread watchlist issues
+			$this->wg->User->clearNotification( $title );
+			$this->wg->Out->setPageTitle( $wallMessage->getMetaTitle() );
 		}
 
 		// TODO: keep the varnish cache and do purging on post
```
Comment (WikiaTech) [inline extensions/wikia/Wall/WallBaseController.class.php:78]: INFO Complete the task associated to this "TODO" comment. [rule](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS1135)

File: extensions/wikia/Wall/WallHooksHelper.class.php [hunk 9]
```diff
@@ -1310,9 +1344,7 @@ static public function onXmlNamespaceSelectorAfterGetFormattedNamespaces( &$name
 	 * @return bool
 	 * @author Andrzej 'nAndy' Łukaszewski
 	 */
-	static public function onWikiaRecentChangesBlockHandlerChangeHeaderBlockGroup( $oChangeList, $r, $oRCCacheEntryArray, &$changeRecentChangesHeader, $oTitle, &$headerTitle ) {
-		wfProfileIn( __METHOD__ );
-
+	static public function onWikiaRecentChangesBlockHandlerChangeHeaderBlockGroup( ChangesList $oChangeList, $r, $oRCCacheEntryArray, &$changeRecentChangesHeader, Title $oTitle, &$headerTitle ) {
```
Comment (WikiaTech) [inline extensions/wikia/Wall/WallHooksHelper.class.php:1347]: MAJOR Remove the unused function parameter "$r". [rule](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS1172)

File: extensions/wikia/Wall/WallHooksHelper.class.php [hunk 10]
```diff
@@ -1491,13 +1531,17 @@ static public function onArticleSaveComplete( &$article, &$user, $text, $summary
...
-	 * @param $editor
+	 * Don't send email for Wall changes
+	 * @param User $editor
 	 * @param Title $title
 	 * @return bool
 	 */
-	static public function onAllowNotifyOnPageChange( $editor, $title ) {
-		$app = F::app();
-		if ( in_array( MWNamespace::getSubject( $title->getNamespace() ), $app->wg->WallNS ) || $title->getNamespace() == NS_USER_WALL_MESSAGE_GREETING ) {
+	static public function onAllowNotifyOnPageChange( User $editor, Title $title ) {
```
Comment (WikiaTech) [inline extensions/wikia/Wall/WallHooksHelper.class.php:1539]: MAJOR Remove the unused function parameter "$editor". [rule](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS1172)

File: extensions/wikia/Wall/WallHooksHelper.class.php [hunk 11]
```diff
@@ -920,20 +947,27 @@ static public function onChangesListInsertArticleLink( $list, &$articleLink, &$s
...
 
-				# Bolden pages watched by this user
-				# Check if the user is following the thread or the board
-				$user = $app->wg->User;
-				if ( $wm->isWatched( $user ) || $wm->isWallWatched( $user ) ) {
-					$articleLink = '<strong class="mw-watched">' . $articleLink . '</strong>';
+				// VOLDEV-3: Bolden pages watched by this user
+				$user = $list->getUser();
+				$specialPageTitle = $list->getTitle();
+				if ( self::shouldBoldenFollowedLink( $user, $specialPageTitle, $wm ) ) {
```
Comment (WikiaTech) [inline extensions/wikia/Wall/WallHooksHelper.class.php:964]: MAJOR Use "static" keyword instead of "self". [rule](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS2037)

File: extensions/wikia/Wall/WallHooksHelper.class.php [hunk 12]
```diff
@@ -920,20 +947,27 @@ static public function onChangesListInsertArticleLink( $list, &$articleLink, &$s
...
+					$articleLink = self::getBoldFollowedLink( $articleLink );
```
Comment (WikiaTech) [inline extensions/wikia/Wall/WallHooksHelper.class.php:965]: MAJOR Use "static" keyword instead of "self". [rule](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS2037)

File: extensions/wikia/Wall/WallHooksHelper.class.php [hunk 13]
```diff
@@ -2286,6 +2330,49 @@ public static function onArticleCommentGetSquidURLs( $title, &$urls ) {
...
 	/**
+	 * SOC-1161: Determine whether a link pointing to a followed thread should be bold
+	 * A link should be bold if it's followed by the user, we are on Special:Watchlist
+	 * and the page was changed since the user's last visit
+	 * @param User $user
+	 * @param Title $title Title of the special page (RecentChanges or Watchlist)
+	 * @param WallMessage $wm
+	 * @return bool Whether to bolden the link
+	 */
+	private static function shouldBoldenFollowedLink( User $user, Title $title, WallMessage $wm ) {
```
Comment (WikiaTech) [inline extensions/wikia/Wall/WallHooksHelper.class.php:2341]: MAJOR Reduce the number of returns of this function 4, down to the maximum allowed 3. [rule](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS1142)

File: extensions/wikia/Wall/WallHooksHelper.class.php [hunk 14]
```diff
@@ -697,10 +697,37 @@ static public function onArticleDelete( $article, &$user, &$reason, &$error ) {
...
+	 * Adjusting Recent Changes for Wall
+	 * Change RC db key to affect Watchlist behavior
+	 * @author TK-999 <https://github.com/TK-999>
+	 * @param RecentChange $rc
+	 * @return Bool true because it is a hook
+	 */
+	static public function onRecentChangeBeforeSave( RecentChange &$rc ) {
+		$wg = F::app()->wg;
+
+		if ( MWNamespace::isTalk( $rc->getAttribute( 'rc_namespace' ) ) &&
```
Comment (WikiaTech) [inline extensions/wikia/Wall/WallHooksHelper.class.php:709]: MINOR Define a constant instead of duplicating this literal "rc_namespace" 25 times. [rule](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS1192)

File: extensions/wikia/Wall/WallHooksHelper.class.php [hunk 15]
```diff
@@ -1324,13 +1356,21 @@ static public function onWikiaRecentChangesBlockHandlerChangeHeaderBlockGroup( $
...
-			$headerTitle = wfMessage( static::getMessagePrefix( $namespace ) . '-thread-group' )
+			$headerTitle =$oChangeList->msg( static::getMessagePrefix( $namespace ) . '-thread-group' )
 				->rawParams( $threadLink )
 				->params( $titleData['wallTitleTxt'], $titleData['wallPageName'] )
 				->parse();
+
+			// VOLDEV-3: Bolden pages watched by this user
+			$user = $oChangeList->getUser();
+			$specialPageTitle = $oChangeList->getTitle();
+			if ( self::shouldBoldenFollowedLink( $user, $specialPageTitle, WallMessage::newFromTitle( $oTitle ) ) ) {
```
Comment (WikiaTech) [inline extensions/wikia/Wall/WallHooksHelper.class.php:1367]: MAJOR Use "static" keyword instead of "self". [rule](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS2037)

File: extensions/wikia/Wall/WallHooksHelper.class.php [hunk 16]
```diff
@@ -1324,13 +1356,21 @@ static public function onWikiaRecentChangesBlockHandlerChangeHeaderBlockGroup( $
...
+				$headerTitle = self::getBoldFollowedLink( $headerTitle );
```
Comment (WikiaTech) [inline extensions/wikia/Wall/WallHooksHelper.class.php:1368]: MAJOR Use "static" keyword instead of "self". [rule](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS2037)

File: extensions/wikia/ArticleComments/classes/ArticleComment.class.php [hunk 17]
```diff
@@ -1127,18 +1127,19 @@ static public function addArticlePageToWatchlist( $comment ) {
...
 		Hooks::run( 'AC_RecentChange_Save', [ &$oRC ] );
 
-		if ( !empty( $wgEnableGroupedArticleCommentsRC ) && ( $oRC instanceof RecentChange ) ) {
-			$title = $oRC->getAttribute( 'rc_title' );
-			$namespace = $oRC->getAttribute( 'rc_namespace' );
-			$article_id = $oRC->getAttribute( 'rc_cur_id' );
-			$title = Title::newFromText( $title, $namespace );
+		if ( $wg->EnableGroupedArticleCommentsRC && ( $oRC instanceof RecentChange ) ) {
+			$title = $oRC->getAttribute('rc_title');
+			$namespace = $oRC->getAttribute('rc_namespace');
```
Comment (WikiaTech) [inline extensions/wikia/ArticleComments/classes/ArticleComment.class.php:1136]: MINOR Define a constant instead of duplicating this literal "rc_namespace" 3 times. [rule](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS1192)

File: extensions/wikia/ArticleComments/classes/ArticleComment.class.php [hunk 18]
```diff
@@ -1127,18 +1127,19 @@ static public function addArticlePageToWatchlist( $comment ) {
...
+			$article_id = $oRC->getAttribute('rc_cur_id');
+			$title = Title::newFromText($title, $namespace);
 
-			// TODO: review
-			if ( MWNamespace::isTalk( $namespace ) &&
+			//TODO: review
```
Comment (WikiaTech) [inline extensions/wikia/ArticleComments/classes/ArticleComment.class.php:1140]: INFO Complete the task associated to this "TODO" comment. [rule](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS1135)

**Comment by WikiaTech** [comment]
SonarQube analysis reported 240 issues
- MAJOR 139 major
- MINOR 75 minor
- INFO 26 info

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:
1. MAJOR [ArticleComment.class.php](https://github.com/Wikia/app/blob/5a54a1c53fe4181ea3e2d44c4db88600538efac4/extensions/wikia/ArticleComments/classes/ArticleComment.class.php): File "ArticleComment.class.php" has 1671 lines, which is greater than 1000 authorized. Split it into smaller files. [rule](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS104)
2. MAJOR [ArticleComment.class.php#L8](https://github.com/Wikia/app/blob/5a54a1c53fe4181ea3e2d44c4db88600538efac4/extensions/wikia/ArticleComments/classes/ArticleComment.class.php#L8): Class "ArticleComment" has 59 methods, which is greater than 20 authorized. Split it into smaller classes. [rule](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS1448)
3. MAJOR [ArticleComment.class.php#L220](https://github.com/Wikia/app/blob/5a54a1c53fe4181ea3e2d44c4db88600538efac4/extensions/wikia/ArticleComments/classes/ArticleComment.class.php#L220): Reduce the number of returns of this function 4, down to the maximum allowed 3. [rule](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS1142)
4. MAJOR [ArticleComment.class.php#L343](https://github.com/Wikia/app/blob/5a54a1c53fe4181ea3e2d44c4db88600538efac4/extensions/wikia/ArticleComments/classes/ArticleComment.class.php#L343): Use "static" keyword instead of "self". [rule](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS2037)
5. MAJOR [ArticleComment.class.php#L437](https://github.com/Wikia/app/blob/5a54a1c53fe4181ea3e2d44c4db88600538efac4/extensions/wikia/ArticleComments/classes/ArticleComment.class.php#L437): Use "static" keyword instead of "self". [rule](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS2037)
6. MAJOR [ArticleComment.class.php#L457](https://github.com/Wikia/app/blob/5a54a1c53fe4181ea3e2d44c4db88600538efac4/extensions/wikia/ArticleComments/classes/ArticleComment.class.php#L457): Use "static" keyword instead of "self". [rule](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS2037)
7. MAJOR [ArticleComment.class.php#L509](https://github.com/Wikia/app/blob/5a54a1c53fe4181ea3e2d44c4db88600538efac4/extensions/wikia/ArticleComments/classes/ArticleComment.class.php#L509): Use "static" keyword instead of "self". [rule](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS2037)
8. MAJOR [ArticleComment.class.php#L510](https://github.com/Wikia/app/blob/5a54a1c53fe4181ea3e2d44c4db88600538efac4/extensions/wikia/ArticleComments/classes/ArticleComment.class.php#L510): Use "static" keyword instead of "self". [rule](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS2037)
9. MAJOR [ArticleComment.class.php#L531](https://github.com/Wikia/app/blob/5a54a1c53fe4181ea3e2d44c4db88600538efac4/extensions/wikia/ArticleComments/classes/ArticleComment.class.php#L531): Use "static" keyword instead of "self". [rule](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS2037)
10. MAJOR [ArticleComment.class.php#L557](https://github.com/Wikia/app/blob/5a54a1c53fe4181ea3e2d44c4db88600538efac4/extensions/wikia/ArticleComments/classes/ArticleComment.class.php#L557): Remove the unused function parameter "$content". [rule](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS1172)

**Review by garthwebb — COMMENTED** [review]
LGTM.  Code needs merge from master and then check unit tests.
